src/protocol/codecs/payload.rs

```diff
@@ -0,0 +1,93 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::Span;
+
+use crate::protocol::{
+    codecs::{
+        tagmsg::Tag,
+        topic::{Topic, TopicCodec},
+    },
+    invalid_data,
+};
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Constructs a [MsgOfInterest] payload from the [topics] list.
+    pub fn construct(mut topics: Vec<Topic>) -> Result<Self, io::Error> {
+        if topics.len() != 1 {
```

Comment:
Well yes and no. We expect a single topic for certain messages that use this topic encoding—nothing else. To ensure the message conforms to the network protocol.

So I would like to fully parse the topics within the message that needs to be constructed with the list of those topics. In this case, the "tags" topic is a single topic for the `MsgOfInterest` message.

Maybe it will make more sense if I move MsgOfInterest's definition to the topic.rs